---
title: "lecture_14_walkthrough"
author: "LQ"
date: "5/17/2019 - Class 14: Transcriptomics and the analysis of RNA-Seq d"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
counts <- read.csv("airway_scaledcounts.csv", stringsAsFactors = FALSE)
metadata <-  read.csv("airway_metadata.csv", stringsAsFactors = FALSE)
counts
metadata
```


##Q1.  How many genes are we dealing with here?
```{r}
nrow(counts)
```


Let's check to see is the format of the labels is correct between the colnames and the counts so that DESeq2 will work correctly and accurately

```{r}
colnames(counts)
metadata$id

```
```{r}
colnames(counts[-1])
```

```{r}
colnames(counts)[-1] == metadata$id
```

```{r}
all(colnames(counts)[-1] == metadata$id)
```

```{r}
control <- metadata[metadata[,"dex"]=="control",]
control.mean <- rowSums( counts[ ,control$id] )/4 
names(control.mean) <- counts$ensgene
```

ONE AT A TIME

```{r}
control <- metadata[metadata[,"dex"]=="control",]
control

```

Now we want to get the cols in counts with these 'control$id' colnames
Changing the last "4" to make the code more robust

```{r}
control.mean <- rowSums( counts[ ,control$id] )/length(control$id)
head(control.mean)
```

###Keeping things organized so assigning the names of the genes to the means

```{r}
names(control.mean) <- counts$ensgene
names(control.mean)
head(control.mean)
```

###Now do the same for the 'treated' samples

```{r}
treated <- metadata[metadata[,"dex"]=="treated",]
treated.mean <- rowSums( counts[ ,treated$id] )/length(treated$id)
names(treated.mean) <- counts$ensgene
names(treated.mean)
head(treated.mean)
```

Let's put these ttogether into a new data.frame called 'meancounts'

```{r}
meancounts <- data.frame(control.mean, treated.mean)
head(meancounts)

```

Plot treated vs control (ybyx)

```{r}
plot(meancounts, xlab = "Control Means", ylab = "Treated Means", col = "mediumvioletred")

```


Q6. Try plotting both axes on a log scale. What is the argument to plot() that allows you to do this? 

```{r}
plot(meancounts, xlab = "Control Means", ylab = "Treated Means", col = "mediumvioletred", log = "xy")

```


```{r}
meancounts$log2fc <- log2(meancounts[,"treated.mean"]/meancounts[,"control.mean"])
head(meancounts)
```

The NaN is returned when you divide by zero and try to take the log. The -Inf is returned when you try to take the log of zero. It turns out that there are a lot of genes with zero expression. Let’s filter our data to remove these genes. Again inspect your result (and the intermediate steps) to see if things make sense to you

```{r}
##Identify the rows (i.e. genes) with zero values
zero.vals <- which(meancounts[,1:2]==0, arr.ind=TRUE)
to.rm <- unique(zero.vals[,1])

##can make the above two lines into one unique(which(meancounts[,1:2]==0, arr.ind = TRUE)[,1])

##Remove these rows as we can't really say much about them
mycounts <- meancounts[-to.rm,]
head(mycounts)
```

##A common threshold used for calling something differentially expressed is a log2(FoldChange) of greater than 2 or less than -2. Let’s filter the dataset both ways to see how many genes are up or down-regulated.
```{r}
up.ind <- mycounts$log2fc > 2
down.ind <- mycounts$log2fc < (-2)
length (which(up.ind == TRUE))
length (which(down.ind == TRUE))

##could also use sum because True = 1 and False = 0; sum(up.ind)
```


##Section 4, Reading annotation data
```{r}
anno <- read.csv("annotables_grch38.csv")
head(anno)
```

We will use the **merge()** function to add annotation data to our 'mycounts' results

```{r}
mycounts.anno <- merge(x = mycounts, y = anno, by.x = "row.names", by.y = "ensgene")
head(mycounts.anno)
```

**BiocManager::install("org.Hs.eg.db")
**library("AnnotationDbi")
**library("org.Hs.eg.db")


```{r}
##BiocManager::install("org.Hs.eg.db")
library("AnnotationDbi")
library("org.Hs.eg.db")

columns(org.Hs.eg.db)

```


##Q11. Run the mapIds() function two more times to add the Entrez ID and UniProt accession as new columns called mycounts$entrez and mycounts$uniprot.
```{r}
mycounts$symbol <- mapIds(org.Hs.eg.db,
                     keys=row.names(mycounts), # Our genenames
                     keytype="ENSEMBL",        # The format of our genenames
                     column="SYMBOL",          # The new format we want to add
                     multiVals="first")
mycounts$entrez <- mapIds(org.Hs.eg.db,
                     keys=row.names(mycounts), # Our genenames
                     keytype="ENSEMBL",        # The format of our genenames
                     column="ENTREZID",          # The new format we want to add
                     multiVals="first")
mycounts$uniprot <- mapIds(org.Hs.eg.db,
                     keys=row.names(mycounts), # Our genenames
                     keytype="ENSEMBL",        # The format of our genenames
                     column="UNIPROT",          # The new format we want to add
                     multiVals="first")
head(mycounts)
```

##Q12. Examine your annotated results for those genes with a log2(FoldChange) of greater than 2 (or less than -2 if you prefer) with the View( mycounts[up.ind,] ) function. What do you notice? Would you trust these results? Why or why not?
```{r}
##View(mycounts[up.ind,])
```

##5. DESeq2 analysis
Let's do this the right way. DESeq2 is an R package for analyzing count-based NGS data like RNA-seq. It is available from Bioconductor. 



```{r}
library(DESeq2)
citation("DESeq2")
```


##Build our DEseq dataset object
```{r}
dds <- DESeqDataSetFromMatrix(countData=counts, 
                              colData=metadata, 
                              design=~dex, 
                              tidy=TRUE)
dds
```


```{r}
dds <- DESeq(dds)
```

```{r}
res <- results(dds)
as.data.frame(res)

```

```{r}
summary(res)
```

###Changing p-value to 0.05
```{r}
res05 <- results(dds, alpha=0.05)
summary(res05)
```


##Let's make a commonly produced visualization from this data, namely a so-called Volcano plot. These summary figures are frequently used to highlight the proportion of genes that are both significantly regulated and display a high fold change.

Typically these plots shows the log fold change on the X-axis, and the −log10 of the p-value on the Y-axis (the more significant the p-value, the larger the −log10 of that value will be). A very dull (i.e. non colored and labeled) version can be created with a quick call to plot() like so:
Without the minus for the y-axis, the volcano plot would be upsidedown so for convention the minus is put in to flip the graph
```{r}
mycols <- rep("gray", nrow(res))
mycols[res$padj<0.05] <- "red"

mycols[abs(res$log2FoldChange)>2] <- "blue"

plot( res$log2FoldChange,  -log(res$padj), 
      xlab="Log2(FoldChange)",
      ylab="-Log(P-value)", col=mycols)
abline(h=-log(0.05), col = "darkgray", lty=2)
abline(v=c(-2,2), col = "darkgray", lty=2)
```

##Another way to plot this is with the Enhanced, but first let's put proper gene names on our results(i.e. gene'symbols')

```{r}
x <- as.data.frame(res)
x$symbol <- mapIds(org.Hs.eg.db, 
                   keys=row.names(x),
                   keytype="ENSEMBL",
                   column="SYMBOL",
                   multiVals="first")

```

```{r}
head(x)
```

```{r}
##STRAIGHT FROM WORKBOOK

# Setup our custom point color vector 
mycols <- rep("gray", nrow(res))
mycols[ abs(res$log2FoldChange) > 2 ]  <- "red" 

inds <- (res$padj < 0.01) & (abs(res$log2FoldChange) > 2 )
mycols[ inds ] <- "blue"

# Volcano plot with custom colors 
plot( res$log2FoldChange,  -log(res$padj), 
 col=mycols, ylab="-Log(P-value)", xlab="Log2(FoldChange)" )

# Cut-off lines
abline(v=c(-2,2), col="gray", lty=2)
abline(h=-log(0.1), col="gray", lty=2)
```

```{r}
##BiocManager::install("EnhancedVolcano")
library(EnhancedVolcano)
EnhancedVolcano(x,
    lab = x$symbol,
    x = 'log2FoldChange',
    y = 'pvalue')
```

















































